sycl/include/CL/sycl/detail/cstr.hpp

```diff
@@ -17,159 +17,144 @@
 #ifndef SYCL_DETAIL_CSTR_HPP
 #define SYCL_DETAIL_CSTR_HPP
 
+#include <algorithm>
+
 #include "CL/sycl/detail/pi.h"
-#include <tuple>
 
 __SYCL_INLINE_NAMESPACE(cl) {
 
-  namespace sycl {
+namespace sycl {
 
-  namespace detail {
+namespace detail {
 
 #ifdef __clang__
-  template <std::size_t I, typename... Ts>
-  using type_at = __type_pack_element<I, Ts...>;
+template <std::size_t I, typename... Ts>
+using type_at = __type_pack_element<I, Ts...>;
 #else
-  template <unsigned I, typename T, typename... Ts> struct type_at_impl {
-    using type = typename type_at_impl<I - 1, Ts...>::type;
-  };
+template <unsigned I, typename T, typename... Ts> struct type_at_impl {
+  using type = typename type_at_impl<I - 1, Ts...>::type;
+};
 
-  template <typename T, typename... Ts> struct type_at_impl<0, T, Ts...> {
-    using type = T;
-  };
+template <typename T, typename... Ts> struct type_at_impl<0, T, Ts...> {
+  using type = T;
+};
 
-  template <unsigned I, typename... Ts>
-  using type_at = typename type_at_impl<I, Ts...>::type;
+template <unsigned I, typename... Ts>
+using type_at = typename type_at_impl<I, Ts...>::type;
 #endif
 
-  /// Utility to make any functor's operator() be const even if it originally
-  /// wasn't. This is useful to bypass the restriction that kernel operator()
-  /// must be const.
-  template<typename T, int Param> struct const_cheating_wrapper {
-    mutable std::decay_t<T> t;
-    template <typename... Ts,
-              typename std::enable_if<sizeof...(Ts) == Param, int>::type = 0>
-    auto operator()(Ts &&...ts) const & {
-      return std::forward<T &>(t)(std::forward<Ts>(ts)...);
-    }
-    template <typename... Ts,
-              typename std::enable_if<sizeof...(Ts) == Param, int>::type = 0>
-    auto operator()(Ts &&...ts) const && {
-      return std::forward<T &&>(t)(std::forward<Ts>(ts)...);
-    }
-  };
-  template<int Param, typename T> auto make_const_cheater(T &&t) {
-    return const_cheating_wrapper<T, Param>{std::forward<T>(t)};
+/// Utility to make any functor's operator() be const even if it originally
+/// wasn't. This is useful to bypass the restriction that kernel operator()
+/// must be const.
```

Comment:
Do we still need this?

Reply:
It is still used when forwarding compile option to v++ through `kernel_param`. 
I'll try to see if we can move kernel_param to the new property system.

Reply:
Yes probably it is a good opportunity to clean this and also introduce a simpler syntax.

Reply:
Removed in d8fac7a